contracts/liquidityMining/KyberRewardLocker.sol

```diff
@@ -0,0 +1,351 @@
+  event SetVestingConfig(IERC20Ext indexed token, uint64 lockTime, uint64 negligibleTimeDifferent);
+
+  /* ========== MODIFIERS ========== */
+
+  modifier onlyRewardsContract(IERC20Ext token) {
+    require(rewardContractsPerToken[token].contains(msg.sender), 'onlyRewardcontract');
+    _;
+  }
+
+  constructor(address _admin) PermissionAdmin(_admin) {}
+
+  /**
+   * @notice Add a whitelisted rewards contract
+   */
+  function addRewardsContract(IERC20Ext token, address _rewardContract) external onlyAdmin {
+    require(
+      rewardContractsPerToken[token].length() < MAX_REWARD_CONTRACTS_SIZE,
+      'rewardContracts is too long'
+    );
+    require(rewardContractsPerToken[token].add(_rewardContract), '_rewardContract is added');
+
+    emit RewardContractAdded(_rewardContract, true);
+  }
+
+  /**
+   * @notice Remove a whitelisted rewards contract
+   */
+  function removeRewardsContract(IERC20Ext token, address _rewardContract) external onlyAdmin {
+    require(rewardContractsPerToken[token].remove(_rewardContract), '_rewardContract is removed');
+
+    emit RewardContractAdded(_rewardContract, false);
+  }
+
+  function setSlashingTarget(IERC20Ext token, address target) external onlyAdmin {
+    slashingTargets[token] = target;
+
+    emit SetSlashingTarget(token, target);
+  }
+
+  function setVestingConfig(
+    IERC20Ext token,
+    uint64 _lockTime,
+    uint64 _negligibleTimeDifferent
+  ) external onlyAdmin {
+    vestingConfigPerToken[token] = VestingConfig({
+      lockTime: _lockTime,
+      negligibleTimeDifferent: _negligibleTimeDifferent
+    });
+
+    emit SetVestingConfig(token, _lockTime, _negligibleTimeDifferent);
+  }
+
+  function lock(
+    IERC20Ext token,
+    address account,
+    uint256 quantity
+  ) external override {
+    lockWithStartTime(token, account, quantity, _blockTimestamp());
+  }
+
+  function lockWithStartTime(
+    IERC20Ext token,
+    address account,
+    uint256 quantity,
+    uint256 startTime
+  ) public override onlyRewardsContract(token) {
+    require(quantity > 0, 'Quantity cannot be zero');
+
+    // transfer token from reward contract to lock contract
+    token.safeTransferFrom(msg.sender, address(this), quantity);
+
+    VestingSchedules storage schedules = accountsVestingSchedules[account][token];
+    uint256 schedulesLength = schedules.length;
+
+    VestingConfig memory config = vestingConfigPerToken[token];
+    uint256 endTime = startTime.add(config.lockTime);
+
+    if (schedulesLength == 0) {
+      accountEscrowedBalance[account][token] = quantity;
+      schedules.data[0] = VestingSchedule({
+        startTime: startTime.toUint64(),
+        endTime: endTime.toUint64(),
+        quantity: quantity.toUint128()
+      });
+      schedules.length = 1;
+    } else {
+      VestingSchedule memory lastSchedule = schedules.data[schedulesLength - 1];
+      uint256 lastLockTime = uint256(lastSchedule.endTime).sub(lastSchedule.startTime);
```

Comment:
maybe change to `lastLockDuration`, got a bit confused bcos of the variable name